File: src/csharp/4_BehavioralPatterns/12_Visitor/Reflective.cs
using System;
using System.Collections.Generic;
using System.Text;
using static System.Console;


namespace DotNetDesignPatternDemos.Behavioral.Visitor.Reflective
{
  using DictType = Dictionary<Type, Action<Expression, StringBuilder>>;

  public abstract class Expression
  {
  }

  public class DoubleExpression : Expression
  {
    public double Value;

    public DoubleExpression(double value)
    {
      Value = value;
    }
  }

  public class AdditionExpression : Expression
  {
    public Expression Left;
    public Expression Right;

    public AdditionExpression(Expression left, Expression right)
    {
      Left = left ?? throw new ArgumentNullException(paramName: nameof(left));
      Right = right ?? throw new ArgumentNullException(paramName: nameof(right));
    }
  }

  

  public static class ExpressionPrinter
  {
    private static DictType actions = new DictType
    {
      [typeof(DoubleExpression)] = (e, sb) =>
      {
        var de = (DoubleExpression) e;
        sb.Append(de.Value);
      },
      [typeof(AdditionExpression)] = (e, sb) =>
      {
        var ae = (AdditionExpression) e;
        sb.Append("(");
        Print(ae.Left, sb);
        sb.Append("+");
        Print(ae.Right, sb);
        sb.Append(")");
      }
    };

    public static void Print2(Expression e, StringBuilder sb)
    {
      actions[e.GetType()](e, sb);
    }

    public static void Print(Expression e, StringBuilder sb)
    {
      if (e is DoubleExpression de)
      {
        sb.Append(de.Value);
      }
      else 
      if (e is AdditionExpression ae)
      {
        sb.Append("(");
        Print(ae.Left, sb);
        sb.Append("+");
        Print(ae.Right, sb);
        sb.Append(")");
      }
      // breaks open-closed principle
      // will work incorrectly on missing case
    }
  }

  public class Demo
  {
    public static void Main()
    {
      var e = new AdditionExpression(
        left: new DoubleExpression(1),
        right: new AdditionExpression(
          left: new DoubleExpression(2),
          right: new DoubleExpression(3)));
      var sb = new StringBuilder();
      ExpressionPrinter.Print2(e, sb);
      WriteLine(sb);
    }
  }
}

File: src/swift/4_BehavioralPatterns/12_Visitor/Reflective.swift
import Foundation
import XCTest

protocol Expression
{

}

class DoubleExpression : Expression
{
  var value: Double
  init(_ value: Double)
  {
    self.value = value
  }
}

class AdditionExpression : Expression
{
  let left, right: Expression

  init(_ left: Expression, _ right: Expression)
  {
    self.left = left
    self.right = right
  }
}

class ExpressionPrinter
{
  // we don't want to use 'self' before initialization
  // so this is marked as 'lazy'
  lazy var actions: [String: (Expression, inout String) -> ()] = [
    "Reflective.DoubleExpression": {
      let de = $0 as! DoubleExpression
      $1.append(String(de.value))
    },
    "Reflective.AdditionExpression": {
      let ae = $0 as! AdditionExpression
      $1.append("(")
      self.print2(ae.left, &$1)
      $1.append("+")
      self.print2(ae.right, &$1)
      $1.append(")")
    }
  ]

  func print(_ e: Expression, _ s: inout String)
  {
    if let de = e as? DoubleExpression
    {
      s.append(String(de.value))
    }
    else if let ae = e as? AdditionExpression
    {
      s.append("(")
      print(ae.left, &s)
      s.append("+")
      print(ae.right, &s)
      s.append(")")
    }
  }

  // breaks open-closed principle
  // will work incorrectly on missing case

  func print2(_ e: Expression, _ s: inout String)
  {
    if let handler = actions[String(describing: e.self)]
    {
      handler(e, &s)
    }
  }
}

func main()
{
  let e = AdditionExpression(
    DoubleExpression(1),
    AdditionExpression(
      DoubleExpression(2),
      DoubleExpression(3)
    )
  )
  var s = ""
  let ep = ExpressionPrinter()
  ep.print(e, &s)
  print(s)

  // try the more concise approach
  s = ""
  ep.print2(e, &s)
  print(s)
}

main()